File: RealtimeTranslator/OpenAI/RealtimeServerErrorClassifier.swift
import Foundation

/// サーバー error イベントの扱い。`shared/fixtures/v1/server-error.json` が正本。
enum RealtimeServerErrorDisposition: Sendable, Equatable {
    /// 接続を維持し、termination も下流イベントも出さない。
    case keepAlive
    /// 既存の再接続へ倒す。
    case recover
    /// 録音を止めて Error にする。
    case halt
}

/// `error.type` / `error.code` を別々に受け取り、許可リストだけで分類する。
/// 判定に使わなかった文字列や生のサーバー文言は保持しない。
struct RealtimeServerErrorClassification: Sendable, Equatable {
    static let transportCode = "transport"

    private static let keepAliveCodes: Set<String> = ["input_audio_buffer_commit_empty"]
    private static let haltCodes: Set<String> = ["insufficient_quota", "billing_hard_limit_reached"]
    private static let recoverCodes: Set<String> = ["server_error", "rate_limit_exceeded", "session_expired"]
    private static let recoverTypes: Set<String> = ["server_error", "rate_limit_error"]

    let disposition: RealtimeServerErrorDisposition
    let termination: EventDeliveryTermination

    static func classify(errorType: String?, code: String?, message: String) -> RealtimeServerErrorClassification {
        classifyCore(errorType: errorType, code: code, message: message, fallback: .halt)
    }

    static func classifyTranscriptionFailure(
        errorType: String?,
        code: String?
    ) -> RealtimeServerErrorClassification {
        classifyCore(errorType: errorType, code: code, message: "", fallback: .keepAlive)
    }

    private static func classifyCore(
        errorType: String?,
        code: String?,
        message: String,
        fallback: RealtimeServerErrorDisposition
    ) -> RealtimeServerErrorClassification {
        let normalizedCode = normalize(code)
        let normalizedType = normalize(errorType)

        // 認証失敗は code に関わらず最優先（transport 扱いで再接続に回さない）。
        // code と type は独立して認証判定に回す（type だけに根拠がある場合も拾う）。
        if RealtimeTranslationError.isAuthenticationFailure(code: code, message: message)
            || RealtimeTranslationError.isAuthenticationFailure(code: errorType, message: message)
        {
            return RealtimeServerErrorClassification(disposition: .halt, termination: .authenticationFailed)
        }

        if normalizedCode == transportCode {
            return RealtimeServerErrorClassification(disposition: .recover, termination: .transportFailure)
        }

        if matches(haltCodes, normalizedCode) || matches(haltCodes, normalizedType) {
            return fatal(message)
        }

        if matches(keepAliveCodes, normalizedCode) {
            return RealtimeServerErrorClassification(disposition: .keepAlive, termination: .none)
        }

        if matches(recoverCodes, normalizedCode) || matches(recoverTypes, normalizedType) {
            return RealtimeServerErrorClassification(disposition: .recover, termination: .recoverableServerError)
        }

        return fallback == .keepAlive
            ? RealtimeServerErrorClassification(disposition: .keepAlive, termination: .none)
            : fatal(message)
    }

    /// keepAlive は例外にならないため呼び出し側で先に除外する。
    func makeError() -> RealtimeTranslationError {
        switch termination {
        case .authenticationFailed:
            return .authenticationFailed
        case .fatalServerError(let message):
            return .fatalServerError(message)
        case .recoverableServerError:
            return .recoverableServerError
        case .transportFailure:
            return .recoverableTransportFailure("server error transport failure")
        case .receiveOverflow:
            return .receiveOverflow
        case .none:
            return .recoverableTransportFailure("keep-alive server error")
        }
    }

    private static func fatal(_ message: String) -> RealtimeServerErrorClassification {
        RealtimeServerErrorClassification(
            disposition: .halt,
            termination: .fatalServerError(RealtimeTranslationError.sanitizedServerMessage(message))
        )
    }

    private static func normalize(_ value: String?) -> String {
        SecretText.normalizeForMatch(value ?? "").replacingOccurrences(of: " ", with: "")
    }

    private static func matches(_ allowlist: Set<String>, _ normalized: String) -> Bool {
        !normalized.isEmpty && allowlist.contains(normalized)
    }
}

File: windows/src/RealtimeTranslator.Core/OpenAI/RealtimeServerErrorClassifier.cs
using System;
using RealtimeTranslator.Core.Realtime;
using RealtimeTranslator.Core.Security;

namespace RealtimeTranslator.Core.OpenAI;

/// <summary>サーバー error イベントの扱い。<c>shared/fixtures/v1/server-error.json</c> が正本。</summary>
public enum RealtimeServerErrorDisposition
{
    /// <summary>接続を維持し、termination も下流イベントも出さない。</summary>
    KeepAlive,

    /// <summary>既存の再接続へ倒す。</summary>
    Recover,

    /// <summary>録音を止めて Error にする。</summary>
    Halt,
}

/// <summary>
/// <c>error.type</c> / <c>error.code</c> を別々に受け取り、許可リストだけで分類する。
/// 判定に使わなかった文字列や生のサーバー文言は保持しない。
/// </summary>
public readonly record struct RealtimeServerErrorClassification(
    RealtimeServerErrorDisposition Disposition,
    EventDeliveryTermination Termination,
    string? SanitizedMessage)
{
    public const string TransportCode = "transport";

    private static readonly string[] KeepAliveCodes =
    [
        "input_audio_buffer_commit_empty",
    ];

    private static readonly string[] HaltCodes =
    [
        "insufficient_quota",
        "billing_hard_limit_reached",
    ];

    private static readonly string[] RecoverCodes =
    [
        "server_error",
        "rate_limit_exceeded",
        "session_expired",
    ];

    private static readonly string[] RecoverTypes =
    [
        "server_error",
        "rate_limit_error",
    ];

    public static RealtimeServerErrorClassification Classify(string? errorType, string? code, string message)
    {
        ArgumentNullException.ThrowIfNull(message);

        var normalizedCode = Normalize(code);
        var normalizedType = Normalize(errorType);

        // 認証失敗は code に関わらず最優先（transport 扱いで再接続に回さない）。
        // code が無い error は type を認証判定へ回す（既存の fallback と同じ範囲を守る）。
        if (RealtimeTranslationException.IsAuthenticationFailure(code ?? errorType, message))
        {
            return new(
                RealtimeServerErrorDisposition.Halt,
                EventDeliveryTermination.AuthenticationFailed,
                null);
        }

        if (normalizedCode == TransportCode)
        {
            return new(
                RealtimeServerErrorDisposition.Recover,
                EventDeliveryTermination.TransportFailure,
                null);
        }

        if (Matches(HaltCodes, normalizedCode) || Matches(HaltCodes, normalizedType))
        {
            return Fatal(message);
        }

        if (Matches(KeepAliveCodes, normalizedCode))
        {
            return new(
                RealtimeServerErrorDisposition.KeepAlive,
                EventDeliveryTermination.None,
                null);
        }

        if (Matches(RecoverCodes, normalizedCode) || Matches(RecoverTypes, normalizedType))
        {
            return new(
                RealtimeServerErrorDisposition.Recover,
                EventDeliveryTermination.RecoverableServerError,
                null);
        }

        return Fatal(message);
    }

    public RealtimeTranslationException ToException() => Termination switch
    {
        EventDeliveryTermination.AuthenticationFailed =>
            new RealtimeTranslationException(RealtimeTranslationErrorKind.AuthenticationFailed),
        EventDeliveryTermination.FatalServerError =>
            new RealtimeTranslationException(RealtimeTranslationErrorKind.FatalServerError, SanitizedMessage),
        EventDeliveryTermination.RecoverableServerError =>
            new RealtimeTranslationException(RealtimeTranslationErrorKind.RecoverableServerError),
        EventDeliveryTermination.TransportFailure =>
            new RealtimeTranslationException(RealtimeTranslationErrorKind.RecoverableTransportFailure),
        _ => throw new InvalidOperationException("keep-alive errors do not produce an exception."),
    };

    private static RealtimeServerErrorClassification Fatal(string message) => new(
        RealtimeServerErrorDisposition.Halt,
        EventDeliveryTermination.FatalServerError,
        RealtimeTranslationException.SanitizeServerMessage(message));

    private static string Normalize(string? value) =>
        SecretText.NormalizeForMatch(value ?? string.Empty).Replace(" ", string.Empty, StringComparison.Ordinal);

    private static bool Matches(string[] allowlist, string normalized) =>
        normalized.Length > 0 && Array.IndexOf(allowlist, normalized) >= 0;
}
